src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/ExtractToCodeBehindCodeActionProvider.cs

```diff
@@ -0,0 +1,109 @@
+
+        override public Task<CommandOrCodeActionContainer> ProvideAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
+        {
+            if (!FileKinds.IsComponent(context.Document.GetFileKind()))
+            {
+                return EmptyResult;
+            }
+
+            var change = new SourceChange(context.Location.AbsoluteIndex, length: 0, newText: string.Empty);
+            var node = context.Document.GetSyntaxTree().Root.LocateOwner(change);
+            if (node is null)
+            {
+                return EmptyResult;
+            }
+
+            node = node.Ancestors().FirstOrDefault(n => n.Kind == SyntaxKind.RazorDirective);
+            if (node == null || !(node is RazorDirectiveSyntax directiveNode))
+            {
+                return EmptyResult;
+            }
+
+            // Make sure we've found a @code or @functions
+            if (directiveNode.DirectiveDescriptor != ComponentCodeDirective.Directive && directiveNode.DirectiveDescriptor != FunctionsDirective.Directive)
+            {
+                return EmptyResult;
+            }
+
+            // No code action if malformed
+            if (node.GetDiagnostics().Any(d => d.Severity == RazorDiagnosticSeverity.Error))
+            {
+                return EmptyResult;
+            }
+
+            var cSharpCodeBlockNode = directiveNode.Body.DescendantNodes().FirstOrDefault(n => n is CSharpCodeBlockSyntax);
+            if (cSharpCodeBlockNode is null)
+            {
+                return EmptyResult;
+            }
+
+            if (HasUnsupportedChildren(cSharpCodeBlockNode))
+            {
+                return EmptyResult;
+            }
+
+            // Do not provide code action if the cursor is inside the code block
+            if (context.Location.AbsoluteIndex > cSharpCodeBlockNode.SpanStart)
+            {
+                return EmptyResult;
+            }
+
+            var actionParams = new ExtractToCodeBehindParams()
+            {
+                Uri = context.Request.TextDocument.Uri,
+                ExtractStart = cSharpCodeBlockNode.Span.Start,
+                ExtractEnd = cSharpCodeBlockNode.Span.End,
+                RemoveStart = directiveNode.Span.Start,
+                RemoveEnd = directiveNode.Span.End
+            };
+            var data = JObject.FromObject(actionParams);
+
+            var resolutionParams = new RazorCodeActionResolutionParams()
+            {
+                Action = LanguageServerConstants.CodeActions.ExtractToCodeBehindAction,
+                Data = data,
+            };
+            var serializedParams = JToken.FromObject(resolutionParams);
+            var arguments = new JArray(serializedParams);
+
+            var container = new List<CommandOrCodeAction>
+            {
+                new Command()
+                {
+                    Title = "Extract code block into backing document",
+                    Name = "razor/runCodeAction",
```

Comment:
Can you extract this out as a constant somewhere?

Reply:
Where do you think the best place to put these is?

Reply:
[`LanguageServerConstants.cs`](https://github.com/dotnet/aspnetcore-tooling/pull/2039/files/e5c98dfd0f1609927614eca912a20acfb6ef20b7#diff-74294135e5847f4e10b726e39f108d21R31)

Reply:
Nevermind, I see you've already added it there 😄